compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1933,28 +1933,92 @@ TR::Register *OMR::X86::TreeEvaluator::integerDualMulEvaluator(TR::Node *node, T
 
 TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR::Register *targetRegister = 0;
-   TR::Node     *firstChild     = node->getFirstChild();
-   TR::Node     *secondChild    = node->getSecondChild();
-   bool          nodeIs64Bit    = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
-   intptrj_t    value;
+   TR::Register *targetRegister  = 0;
+   TR::Node     *firstChild      = node->getFirstChild();
+   TR::Node     *secondChild     = node->getSecondChild();
+   bool          nodeIs64Bit     = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   bool          isLoadConst     = secondChild->getOpCode().isLoadConst();
+   bool          isByteOperation = false;
+   bool          isMemRef        = false;
+   TR_X86OpCodes opCode          = BADIA32Op; 
+   TR_X86OpCodes regOpCode       = BADIA32Op; 
+   TR_X86OpCodes memOpCode       = BADIA32Op; 
+   TR_X86OpCodes movOpCode       = BADIA32Op;
+   intptrj_t     value;
 
    if (node->isDualCyclic())
       {
       return TR::TreeEvaluator::integerDualMulEvaluator(node, cg);
       }
 
-   if (secondChild->getOpCode().isLoadConst() &&
-      (value = TR::TreeEvaluator::integerConstNodeValue(secondChild, cg)))
+   // Opcode evaluation for bmul, smul imul and lmul
+   switch (secondChild->getDataType())
+      {
+      case TR::Int8:
+         if(isLoadConst) 
+            value = secondChild->getByte();
+         isMemRef = secondChild->getOpCode().isMemoryReference(); 
+         opCode = isMemRef ? IMUL1AccMem : IMUL1AccReg;
+     
+         regOpCode = IMUL1AccReg;
+         memOpCode = IMUL1AccMem;
+         movOpCode = MOV1RegReg;
+
+         isByteOperation = true;
+         break;
+
+      case TR::Int16:
+         if(isLoadConst)
+            {
+            value = secondChild->getShortInt();
+            isMemRef = firstChild->getOpCode().isMemoryReference(); 
+            if (isMemRef)
+               opCode = (value >= -128 && value <= 127) ? IMUL2RegMemImms : IMUL2RegMemImm2;
+            else
+               opCode = (value >= -128 && value <= 127) ? IMUL2RegRegImms : IMUL2RegRegImm2;
+            }
+
+         regOpCode = IMUL2RegReg;
+         memOpCode = IMUL2RegMem;
+         movOpCode = MOV2RegReg;
+         break;
+
+      case TR::Int32:
+      case TR::Int64:
+         if(isLoadConst)
+            {
+            value = TR::TreeEvaluator::integerConstNodeValue(secondChild, cg);
+            //In case of long we cannot handle larger than 4 bytes to be handled in a 4 byte Immediate
+            TR_ASSERT_FATAL(sizeof(value) <= 4, "Unexpected size for the node which is greater than 4");
```

Comment:
This isn't gong to do what you want. `sizeof(value)` is the numer of bytes allocated for value, not the size of the value from the constant. I don't think you want that and fatal asserting when a 64bit const doesn't fit in 4 bytes is not an acceptable behavior - the codegen needs to handle this properly.